etc/systemd/system/zfs-volume@.service.in

```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=ZFS zvol %I ready
+Before=zfs-volumes.target
+DefaultDependencies=no
+
+[Service]
+ExecStart=/bin/sh -c 'timeout=30 ; while [ $timeout -gt 0 ] ; do [ -e "/dev/zvol/$0" ] && exit 0 ; sleep 1 ; timeout=$((timeout-1)); done ; echo "zvol $0 unavailable"' %I
```

Comment:
this is systemd, no need for such complicated construct :-)

~~~
BindsTo=%i.device
After=%i.device
Conflicts=shutdown.target  # maybe not needed, needs testing
~~~

Reply:
Wow, that's a lot nicer!

I'm not super familiar with how the timeout on this works. If this device is slow to show up, will this give up and enter a failed state, or forever remain in a waiting state? Is this what `JobTimeoutSec=` does (if so, when does it become "queued"?) The other timeout setting I see in `man systemd.unit` seems to do with once the job has fired.  Is my understanding correct?

Reply:
Sorry for my late response, I Will have a look again later rhis day

Reply:
JobTimeoutSec is there for a) setting a timeout other than the default (infinity) for normal units and b) for overriding the default device unit timeout (which is specified by DefaultTimeoutStartSec=). In our case b) applies, which we should not modify upstream. DefaultTimeoutStartSec is 90s for Debian for example.

When the timeout is reached, the unit gets aborted and enters "failed" state. What happens then depends on the dependencies specified by "Wants" and "Requires". When the zvol target requires all zvol@.service and the zvol target is required by multi-user.target, then systemd will completely refuse to boot and go to recovery mode. This is *not* what we want but since zfs-volumes.target is only WantedBy=zfs.target, it only gets marked as failed.